presto-parser/src/main/java/com/facebook/presto/sql/tree/TemplateExpressionRewriter.java

```diff
@@ -0,0 +1,112 @@
+
+public class TemplateExpressionRewriter
+{
+    private TemplateExpressionRewriter() {}
+
+    public static Expression rewrite(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private int counter;
+
+        private Visitor()
+        {
+            this.counter = 0;
+        }
+
+        @Override
+        public Expression rewriteBinaryLiteral(BinaryLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteBooleanLiteral(BooleanLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteCharLiteral(CharLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteDecimalLiteral(DecimalLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteDoubleLiteral(DoubleLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteGenericLiteral(GenericLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteIntervalLiteral(IntervalLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteLongLiteral(LongLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteStringLiteral(StringLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteTimeLiteral(TimeLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteTimestampLiteral(TimestampLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        @Override
+        public Expression rewriteNullLiteral(NullLiteral node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            return getParameter();
+        }
+
+        private Parameter getParameter()
+        {
+            return new Parameter(counter++);
```

Comment:
What happens if there is already "Parameter" in this expression? Then there could end up being two parameters with the same index in the expression, and that could cause bugs elsewhere?

Reply:
Updated to replace existing parameters with a parameters constructed with the correct count.